Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -477,6 +477,51 @@ void TextureCacheBase::BindTextures()
   }
 }
 
+class ArbitraryMipmapDetector
+{
+public:
+  void AddLevel(u32 width, u32 height, u32 row_length, const u8* buffer)
+  {
+    ++levels;
+
+    std::array<u8, 4> avg = { 0 };
+    for (u32 i = 0; i < height; ++i)
+    {
+      const auto* row = buffer;
+      for (u32 j = 0; j < width; ++j)
+      {
+        avg[0] += row[0];
+        avg[1] += row[1];
+        avg[2] += row[2];
+        avg[3] += row[3];
+
+        row += 4;
+      }
+      
+      buffer += row_length;
+    }
+
+    for (size_t c = 0; c < avg.size(); ++c)
+    {
+      u8 cur = avg[c] / (width * height);
+      max[c] = std::max(cur, max[c]);
+      min[c] = std::min(cur, min[c]);
+    }
+  }
+
+  bool HasArbitraryMipmaps() const {
+    const u8 THRESHOLD = 50;
+    return
+      levels > 1 &&
+      max[0] - min[0] + max[1] - min[1] + max[2] - min[2] + max[3] - min[3] > THRESHOLD;
+  }
+
+private:
+  u32 levels = 0;
+  std::array<u8, 4> min = { 255 };
+  std::array<u8, 4> max = { 0 };
```

Comment:
The convention we use for zeroed out std::arrays is

```cpp
std::array<type, size> name{};
```

Reply:
I actually didn't know that that was well defined!